File: boss-2026/coin-selection-challenge-winterrdog/solution/cpp/coin-selection.cc
#include "./coin-selection.hh"

void make_payments(const std::vector<Payment>& payments) {
  // Make payments
  for (const Payment& payment : payments) {
    // Get all UTXO in wallet coin pool
    std::string list_unspent{bcli("listunspent")};

    // Objectify coins to spend
    json coins_json{json::parse(list_unspent)};
    std::vector<Coin> coins{coins_json.get<std::vector<Coin>>()};

    // Select coins

    // Create transaction
    std::string inputs{};
    std::string outputs{};
    std::string unsigned_tx{
        bcli("createrawtransaction " + inputs + " " + outputs)};

    // Sign transaction
    std::string signed_tx_res{
        bcli("signrawtransactionwithwallet " + unsigned_tx)};
    json signed_tx_json{json::parse(signed_tx_res)};
    std::string signed_tx{signed_tx_json["hex"].get<std::string>()};

    // Broadcast transaction without maxfeerate protection
    bcli("sendrawtransaction " + signed_tx + " 0");

    // Confirm transaciton
    bcli("generatetoaddress 1 bcrt1pqqqqrldkrl");
  }
}

File: boss-2026/coin-selection-challenge-winterrdog/solution/python/coin-selection.py
#!/usr/bin/env python3

from decimal import Decimal
from subprocess import run
import json
from pathlib import Path

BTC = Decimal("0.00000000")
DESCRIPTOR_FILE = Path(__file__).resolve().parent.parent.parent / "datadir" / "student_wallet.json"
PAYMENTS_FILE = Path(__file__).resolve().parent.parent.parent / "payments.json"

def bcli(cmd: str):
    res = run(
            ["bitcoin-cli", "-regtest", "-rpcuser=student", "-rpcpassword=boss2026"] + cmd.split(" "),
            capture_output=True,
            encoding="utf-8")
    if res.returncode == 0:
        return res.stdout.strip()
    else:
        raise Exception(res.stderr.strip())


# Initialize wallet
bcli("createwallet student false true")
with open(DESCRIPTOR_FILE) as f:
    bcli(f"importdescriptors {f.read().replace(' ','')}")


# Load payments
with open(PAYMENTS_FILE) as f:
    payments = json.loads(f.read())

# Make payments
for payment in payments:
    # Get all UTXO in wallet coin pool
    coins = json.loads(bcli("listunspent"), parse_float=Decimal)

    # Select coins



    # Create transaction
    inputs = []
    outputs = []
    unsigned_tx = bcli(f"createrawtransaction {json.dumps(inputs).replace(' ','')} {json.dumps(outputs).replace(' ','')}")
    # Sign transaction
    signed_tx = json.loads(bcli(f"signrawtransactionwithwallet {unsigned_tx}"))["hex"]
    # Broadcast transaction without maxfeerate protection
    bcli(f"sendrawtransaction {signed_tx} 0")
    # Confirm transaciton
    bcli("generatetoaddress 1 bcrt1pqqqqrldkrl")
